File: 381. Insert Delete GetRandom O(1) - Duplicates allowed/Java/Solution.java
public class RandomizedCollection {
    List<Integer> nums;
    Map<Integer, Set<Integer>> memory;
    Random random;

    /**
     * Initialize your data structure here.
     */
    public RandomizedCollection() {
        nums = new ArrayList<>();
        memory = new HashMap<>();
        random = new Random();
    }

    /**
     * Inserts a value to the collection. Returns true if the collection did not already contain the specified element.
     */
    public boolean insert(int val) {
        boolean isAlreadyIn = true;
        if (!memory.containsKey(val)) {
            isAlreadyIn = false;
            memory.put(val, new HashSet<>());
        }
        memory.get(val).add(nums.size());
        nums.add(val);
        return !isAlreadyIn;
    }

    /**
     * Removes a value from the collection. Returns true if the collection contained the specified element.
     */
    public boolean remove(int val) {
        if (!memory.containsKey(val)) {
            return false;
        }
        int lastIndex = nums.size() - 1;
        if (!memory.get(val).contains(lastIndex)) {
            int currentIndex = memory.get(val).iterator().next();
            int lastValue = nums.get(nums.size() - 1);
            memory.get(lastValue).remove(lastIndex);
            memory.get(lastValue).add(currentIndex);
            memory.get(val).remove(currentIndex);
            nums.set(currentIndex, lastValue);
        } else {
            memory.get(val).remove(lastIndex);
        }
        if (memory.get(val).isEmpty()) {
            memory.remove(val);
        }
        nums.remove(nums.size() - 1);
        return true;
    }

    /**
     * Get a random element from the collection.
     */
    public int getRandom() {
        return nums.get(random.nextInt(nums.size()));
    }
}

/**
 * Your RandomizedCollection object will be instantiated and called as such:
 * RandomizedCollection obj = new RandomizedCollection();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: 381. Insert Delete GetRandom O(1) - Duplicates allowed/Python/Solution.py
class RandomizedCollection(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.nums = []
        self.memory = {}

    def insert(self, val):
        """
        Inserts a value to the collection. Returns true if the collection did not already contain the specified element.
        :type val: int
        :rtype: bool
        """
        is_already_in = True
        if val not in self.memory:
            is_already_in = False
            self.memory[val] = set()
        self.memory[val].add(len(self.nums))
        self.nums.append(val)
        return not is_already_in

    def remove(self, val):
        """
        Removes a value from the collection. Returns true if the collection contained the specified element.
        :type val: int
        :rtype: bool
        """
        if val not in self.memory:
            return False
        last_index = len(self.nums) - 1
        if last_index not in self.memory[val]:
            current_index = next(iter(self.memory[val]))
            last_value = self.nums[-1]
            self.memory[last_value].remove(last_index)
            self.memory[last_value].add(current_index)
            self.memory[val].remove(current_index)
            self.nums[current_index] = last_value
        else:
            self.memory[val].remove(last_index)
        if not self.memory[val]:
            del self.memory[val]
        self.nums.pop()
        return True

    def getRandom(self):
        """
        Get a random element from the collection.
        :rtype: int
        """
        return random.choice(self.nums)


# Your RandomizedCollection object will be instantiated and called as such:
# obj = RandomizedCollection()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
